sponsorship_compassion/model/contracts.py

```diff
@@ -392,22 +378,393 @@ def contract_waiting_mandate(self, cr, uid, ids, context=None):
+            inv_update_ids = set()
+            for contract in self.browse(cr, uid, ids, context):
+                # If some invoices are left cancelled, we update them
+                # with new contract information and validate them
+                cancel_ids = invoice_obj.search(cr, uid, [
+                    ('state', '=', 'cancel'),
+                    ('id', 'in', list(invoices_canceled)),
+                    ('date_invoice', '>=', since_date)], context=context)
+                if cancel_ids:
+                    inv_update_ids.update(cancel_ids)
+                    invoice_obj.action_cancel_draft(cr, uid, cancel_ids)
+                    self._update_invoice_lines(cr, uid, contract, cancel_ids,
+                                               context)
+                # If no invoices are left in cancel state, we rewind
+                # the next_invoice_date for the contract to generate again
+                else:
+                    next_invoice_date = datetime.strptime(
+                        contract.next_invoice_date, DF)
+                    next_invoice_date = next_invoice_date + relativedelta(
+                        months=-len(invoices_canceled))
+                    super(recurring_contract, self).write(
+                        cr, uid, contract.id, {
+                            'next_invoice_date': next_invoice_date.strftime(
+                                DF)}, context)
+                    invoicer_id = contract.group_id.generate_invoices()
+                    invoicer = self.pool.get('recurring.invoicer').browse(
+                        cr, uid, invoicer_id, context)
+                    if invoicer.invoice_ids:
+                        invoicer.validate_invoices()
+                    else:
+                        invoicer.unlink()
+            # Validate again modified invoices
+            if inv_update_ids:
+                wf_service = netsvc.LocalService('workflow')
+                for invoice_id in inv_update_ids:
+                    wf_service.trg_validate(
+                        uid, 'account.invoice', invoice_id,
+                        'invoice_open', cr)
+        return True
+
+    def _on_change_next_invoice_date(self, cr, uid, ids, next_invoice_date,
+                                     context=None):
+        """Prevent to change next_invoice_date in the past
+        for active contract.
+        """
+        for contract in self.browse(cr, uid, ids, context):
+            if contract.state not in ('draft', 'mandate') \
+                    and next_invoice_date < contract.next_invoice_date:
+                raise orm.except_orm(
+                    _('Warning'),
+                    _('You cannot rewind the next invoice date.'))
+
+    def _on_change_child_id(self, cr, uid, ids, child_id, partner_id=None,
+                            context=None):
+        """Link/unlink child to sponsor
+        """
+        for contract in self.browse(cr, uid, ids, context):

```

Comment:
It's good  to write each function to be able to handle multi-ids. But in this case, I think the function should raise an exception if len(ids) != 1. Otherwise, the child->sponsor_id could be inconsistent with contract->child_id.